core/migration.go

```diff
@@ -150,6 +162,43 @@ func (m *Migration) Run(context *Context) {
 	}
 }
 
+func (m *Migration) addSubIds(context *Context, task *AddSubIds) {
+	users := context.GetUsersInGroup(task.GroupName)
+	users = append(users, context.GetRootUser())
+
+	if _, err := os.Stat("/etc/subuid"); err != nil {
```

Comment:
If what you want is to ensure a file exists, this if-ladder could be stripped down a bit, plus making the file check and creation atomic:

```go
_, err := os.OpenFile(os.O_WRONLY|os.O_APPEND|os.O_EXCL)
if err != nil  {
    if !errors.Is(err, fs.ErrExist) {
        fmt.Println("Houston, ...")
        // If Go allows it, you could also use a `goto` to move past this /etc/subuid block.
        // This way you keep well-separated happy and sad paths by avoiding the `else` below.
    }
    fmt.Println("It's all good, this file already existed") // Are we interested in printing this message anyway?
}
// If we're here, either the file exists, or we successully created it. Either way, we're good.
```

To be 100% honest I didn't test this snippet, but please give it a try.

Reply:
I believe we only want to add the sub IDs if the file didn't already exist. If the file does exist, we want to do nothing.

Reply:
@silkeh Would this code below be acceptable regarding nesting? I think this incorporates both concerns:

```go
func createFileIfNotExists(path string) (created bool) {
	fh, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)

	if err != nil {
		if errors.Is(err, os.ErrExist) {
			waterlog.Debugf("\t%s already exists, skipping", path)
		} else {
			waterlog.Warnf("\tUnable to create %s due to error: %s\n", path, err)
		}

		return false
	}

	fh.Close()

	return true
}
```

Reply:
We could also simply return the error and let the caller handle it.

Reply:
> @silkeh Would this code below be acceptable regarding nesting? I think this incorporates both concerns:

Yes, looks fine!

> We could also simply return the error and let the caller handle it.

Yep, that would also work (if it returns `nil` if the error is `os.ErrExist` :slightly_smiling_face:)!